File: bstSearch.c
#include<stdio.h>
typedef struct treeNode
{
	int data;
	struct treeNode *right;
	struct treeNode *left;
}treeNode;

treeNode *root = NULL;
treeNode *temp = NULL;


void search(treeNode *t)
{
	if(t->data > temp->data && t->left != NULL)
		search(t->left);
	else if(t->data > temp->data && t->left == NULL)
		t->left = temp;
	else if(t->data < temp->data && t->right != NULL)
		search(t->right);
	else if(t->data < temp->data && t->right == NULL)
		t->right = temp;
}

void create(int n)
{
	int i;
	for(i = 1; i <= n; i++)
	{
		temp = (treeNode*)malloc(sizeof(treeNode));
		printf("Enter a number ");
		scanf("%d", &temp->data);
		temp->left = NULL;
		temp->right = NULL;
		
		if(root == NULL)
			root = temp;
		else
			search(root);
	}
}

treeNode *searchKey(treeNode *t, int key)
{
	if(t == NULL)
		return NULL;
	if(t->data == key)
		return t;
	else if(t->data > key)
		searchKey(t->left, key);
	else if(t->data < key)
		searchKey(t->right, key);
}

void traverse(treeNode *t)
{
    if (root == NULL)
    {
        printf("No elements in a tree to display");
        return;
    }
    if (t->left != NULL)    
        traverse(t->left);
    printf("%d -> ", t->data);
    if (t->right != NULL)    
        traverse(t->right);
}

int main()
{
	int n;
	printf("Enter a number of nodes u want to create ");
	scanf("%d", &n);
	create(n);
	traverse(root);
	int key;
	printf("Enter the number u want to search ");
	scanf("%d", &key);
	treeNode *found = searchKey(root, key);
	if(found == NULL)
		printf("Value is not present ");
	else
		printf("value is %d ", found->data);
	return 0;
}

File: bstSearch.java

import java.util.*;


class Node
{
    int data;
    Node right;
    Node left;

    Node(int data)
    {
        this.data = data;
        this.right = null;
        this.left = null;
    }
}

public class bstSearch
{
    //root node of the tree
    Node root;

    //function for searching the proper position for inserting a node into a tree
    public void search(Node temp, Node newNode)
    {
        //traversig left side of a tree
        if(temp.data > newNode.data && temp.left != null)
            search(temp.left, newNode);
        if(temp.data > newNode.data && temp.left == null)
            temp.left = newNode;

        //traversign right side of a tree
        if(temp.data < newNode.data && temp.right != null)
            search(temp.right, newNode);
        if(temp.data < newNode.data && temp.right == null)
            temp.right = newNode;
    }

    //function for creating node for a tree and inserting in a tree
    public void createNode(int num)
    {
        Node newNode = new Node(num);
        if(root == null)
            root = newNode;
        else
        {
            Node temp = root;
            search(temp, newNode);
        }
    }

    //traversing a tree in a pre-order manner
    public void traverse(Node temp)
    {
        if(temp == null)
            System.out.println("Tree is empty : ");
        else
        {
            System.out.println(temp.data +  " ");
            if(temp.left != null)
                traverse(temp.left);
            if(temp.right != null)
                traverse(temp.right);
        }
    }

    static boolean iterativeSearch(Node root, int key) 
    { 
        // Traverse untill root reaches to dead end 
        while (root != null) { 
            // pass right subtree as new tree 
            if (key > root.data) 
                root = root.right; 
  
            // pass left subtree as new tree 
            else if (key < root.data) 
                root = root.left; 
            else
                return true; // if the key is found return 1 
        } 
        return false; 
    } 
 

    public static void main(String[] args)
    {
        bstSearch obj = new bstSearch();
        Scanner scan = new Scanner(System.in);
        int n;
        int num;

        System.out.println("Enter the number of nodes in a tree : ");
        n = scan.nextInt();

        for(int i = 0; i < n; i++)
        {
            System.out.println("Enter a number : ");
            num = scan.nextInt();
            obj.createNode(num);
        }
        obj.traverse(obj.root);

        System.out.println("Enter a number that u want to search : ");
        num = scan.nextInt();

        if (iterativeSearch(obj.root, num)) 
            System.out.println("YES"); 
        else
            System.out.println("NO"); 
        scan.close();
    }
}